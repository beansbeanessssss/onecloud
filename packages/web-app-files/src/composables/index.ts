export * from './actions'
export * from './resourcesViewDefaults'
export * from './shares'
