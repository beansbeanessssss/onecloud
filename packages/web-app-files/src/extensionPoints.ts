import { ExtensionPoint } from '@ownclouders/web-pkg'
import { computed } from 'vue'

export const uploadMenuExtensionPoint: ExtensionPoint = {
  id: 'app.files.upload-menu',
  extensionType: 'action',
  multiple: true
}
export const quickActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.quick-actions',
  extensionType: 'action',
  multiple: true
}
export const batchActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.batch-actions',
  extensionType: 'action',
  multiple: true
}
export const contextActionsExtensionPoint: ExtensionPoint = {
  id: 'app.files.context-actions',
  extensionType: 'action',
  multiple: true
}
export const fileSideBarExtensionPoint: ExtensionPoint = {
  id: 'app.files.sidebar',
  extensionType: 'sidebarPanel',
  multiple: true
}
export const folderViewsFolderExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.folder',
  extensionType: 'folderView'
}
export const folderViewsFavoritesExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.favorites',
  extensionType: 'folderView'
}
export const folderViewsProjectSpacesExtensionPoint: ExtensionPoint = {
  id: 'app.files.folder-views.project-spaces',
  extensionType: 'folderView'
}

export const extensionPoints = () => {
  return computed<ExtensionPoint[]>(() => {
    return [
      uploadMenuExtensionPoint,
      quickActionsExtensionPoint,
      batchActionsExtensionPoint,
      contextActionsExtensionPoint,
      fileSideBarExtensionPoint,
      folderViewsFolderExtensionPoint,
      folderViewsFavoritesExtensionPoint,
      folderViewsProjectSpacesExtensionPoint
    ]
  })
}
