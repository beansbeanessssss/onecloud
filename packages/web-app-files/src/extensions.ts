import {
  ApplicationSetupOptions,
  Extension,
  useStore,
  useRouter,
  useSearch,
  useFileActionsShowShares,
  useFileActionsCreateQuickLink
} from '@ownclouders/web-pkg'
import { computed, unref } from 'vue'
import { SDKSearch } from './search'

export const extensions = ({ applicationConfig }: ApplicationSetupOptions) => {
  const store = useStore()
  const router = useRouter()
  const { search: searchFunction } = useSearch()

  const { actions: showSharesActions } = useFileActionsShowShares()
  const { actions: quickLinkActions } = useFileActionsCreateQuickLink()

  return computed(
    () =>
      [
        {
          id: 'com.github.owncloud.web.files.search',
          type: 'search',
          searchProvider: new SDKSearch(store, router, searchFunction)
        },
        {
          id: 'com.github.owncloud.web.files.quick-action.collaborator',
          scopes: ['files', 'files.quick-action'],
          type: 'action',
          action: unref(showSharesActions)[0]
        },
        {
          id: 'com.github.owncloud.web.files.quick-action.quicklink',
          scopes: ['files', 'files.quick-action'],
          type: 'action',
          action: unref(quickLinkActions)[0]
        }
      ] satisfies Extension[]
  )
}
