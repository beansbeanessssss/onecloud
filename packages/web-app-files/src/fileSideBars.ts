import FileDetails from './components/SideBar/Details/FileDetails.vue'
import FileDetailsMultiple from './components/SideBar/Details/FileDetailsMultiple.vue'
import FileActions from './components/SideBar/Actions/FileActions.vue'
import FileVersions from './components/SideBar/Versions/FileVersions.vue'
import SharesPanel from './components/SideBar/Shares/SharesPanel.vue'
import NoSelection from './components/SideBar/NoSelection.vue'
import SpaceActions from './components/SideBar/Actions/SpaceActions.vue'
import {
  SpaceDetails,
  SpaceDetailsMultiple,
  SpaceNoSelection,
  isLocationTrashActive,
  isLocationPublicActive,
  isLocationSpacesActive,
  isLocationSharesActive,
  useCapabilityFilesSharing,
  useCapabilityFilesSharingApiEnabled,
  useCapabilityFilesSharingPublicEnabled,
  useRouter,
  useStore,
  SidebarPanelExtension,
  useIsFilesAppActive
} from '@ownclouders/web-pkg'
import {
  isProjectSpaceResource,
  spaceRoleEditor,
  spaceRoleManager
} from '@ownclouders/web-client/src/helpers'
import { Resource } from '@ownclouders/web-client'
import { useGettext } from 'vue3-gettext'
import { computed, unref } from 'vue'

export const sideBarPanels = () => {
  const router = useRouter()
  const store = useStore()
  const user = computed(() => store.getters.user)
  const { $gettext } = useGettext()
  const isSharingEnabled = useCapabilityFilesSharing(store)
  const isSharingApiEnabled = useCapabilityFilesSharingApiEnabled(store)
  const arePublicLinksEnabled = useCapabilityFilesSharingPublicEnabled(store)
  const isFilesAppActive = useIsFilesAppActive()

  return computed(
    () =>
      [
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.no-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'no-selection',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: NoSelection,
            isRoot: () => true,
            isEnabled: ({ parent, items }) => {
              if (isLocationSpacesActive(router, 'files-spaces-projects')) {
                // project spaces overview has its own "no selection" panel
                return false
              }
              if (items?.length > 0) {
                return false
              }
              // empty selection in a project space root shows a "details" panel for the space itself
              return !isProjectSpaceResource(parent)
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.details-single-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: FileDetails,
            componentAttrs: () => ({
              previewEnabled: unref(isFilesAppActive)
            }),
            isRoot: () => !isLocationTrashActive(router, 'files-trash-generic'),
            isEnabled: ({ items }) => {
              if (isLocationTrashActive(router, 'files-trash-generic')) {
                // details panel is not available in trash
                return false
              }
              if (items?.length !== 1) {
                return false
              }
              // project spaces have their own "details" panel
              return !isProjectSpaceResource(items[0])
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.details-multi-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details-multiple',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: FileDetailsMultiple,
            componentAttrs: () => ({
              get showSpaceCount() {
                return (
                  !isLocationSpacesActive(router, 'files-spaces-generic') &&
                  !isLocationSharesActive(router, 'files-shares-with-me') &&
                  !isLocationTrashActive(router, 'files-trash-generic')
                )
              }
            }),
            isRoot: () => true,
            isEnabled: ({ items }) => {
              if (isLocationSpacesActive(router, 'files-spaces-projects')) {
                // project spaces overview has its own "no selection" panel
                return false
              }
              return items?.length > 1
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.actions',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'actions',
            icon: 'slideshow-3',
            title: () => $gettext('Actions'),
            component: FileActions,
            isRoot: () => isLocationTrashActive(router, 'files-trash-generic'),
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              // project spaces have their own "actions" panel
              return !isProjectSpaceResource(items[0])
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.sharing',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'sharing',
            icon: 'user-add',
            iconFillType: 'line',
            title: () => $gettext('Shares'),
            component: SharesPanel,
            componentAttrs: () => ({
              showSpaceMembers: false,
              get showLinks() {
                if (unref(isSharingEnabled)) {
                  return unref(arePublicLinksEnabled)
                }
                return false
              }
            }),
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              if (isProjectSpaceResource(items[0])) {
                // project space roots have their own "sharing" panel (= space members)
                return false
              }
              if (
                isLocationTrashActive(router, 'files-trash-generic') ||
                isLocationPublicActive(router, 'files-public-link')
              ) {
                return false
              }
              if (unref(isSharingEnabled)) {
                return unref(isSharingApiEnabled)
              }
              return false
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.versions',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'versions',
            icon: 'git-branch',
            title: () => $gettext('Versions'),
            component: FileVersions,
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              if (isProjectSpaceResource(items[0])) {
                // project space roots don't support versions
                return false
              }
              if (
                isLocationTrashActive(router, 'files-trash-generic') ||
                isLocationPublicActive(router, 'files-public-link')
              ) {
                return false
              }
              return items[0].type !== 'folder'
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.no-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'no-selection',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: SpaceNoSelection,
            isRoot: () => true,
            isEnabled: ({ items }) => {
              if (!isLocationSpacesActive(router, 'files-spaces-projects')) {
                // only for project spaces overview
                return false
              }
              return items?.length === 0
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.details-single-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details-space',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: SpaceDetails,
            isRoot: () => true,
            isEnabled: ({ items }) => {
              return items?.length === 1 && isProjectSpaceResource(items[0])
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.details-multi-selection',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'details-space-multiple',
            icon: 'questionnaire-line',
            title: () => $gettext('Details'),
            component: SpaceDetailsMultiple,
            componentAttrs: ({ items }) => ({
              selectedSpaces: items
            }),
            isRoot: () => true,
            isEnabled: ({ items }) => {
              return items?.length > 1 && isLocationSpacesActive(router, 'files-spaces-projects')
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.actions',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'space-actions',
            icon: 'slideshow-3',
            title: () => $gettext('Actions'),
            component: SpaceActions,
            isEnabled: ({ items }) => {
              if (items?.length !== 1) {
                return false
              }
              if (!isProjectSpaceResource(items[0])) {
                return false
              }
              if (
                !isLocationSpacesActive(router, 'files-spaces-projects') &&
                !isLocationSpacesActive(router, 'files-spaces-generic')
              ) {
                return false
              }
              return [
                ...items[0].spaceRoles[spaceRoleManager.name],
                ...items[0].spaceRoles[spaceRoleEditor.name]
              ].some((role) => role.id === unref(user).uuid)
            }
          }
        },
        {
          id: 'com.github.owncloud.web.files.sidebarPanels.projects.sharing',
          type: 'sidebarPanel',
          scopes: ['resource'],
          panel: {
            name: 'space-share',
            icon: 'group',
            title: () => $gettext('Members'),
            component: SharesPanel,
            componentAttrs: () => ({
              showSpaceMembers: true,
              get showLinks() {
                if (unref(isSharingEnabled)) {
                  return unref(arePublicLinksEnabled)
                }
                return false
              }
            }),
            isEnabled: ({ items }) => {
              return items?.length === 1 && isProjectSpaceResource(items[0])
            }
          }
        }
      ] satisfies SidebarPanelExtension<Resource, Resource>[]
  )
}
