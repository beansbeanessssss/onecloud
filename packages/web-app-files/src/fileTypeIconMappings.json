{
  "7z": "file-zip",
  "accdb": "file",
  "ai": "image",
  "apk": "file-zip",
  "bash": "file-code",
  "bz2": "file-zip",
  "cb7": "file-text",
  "cba": "file-text",
  "cbr": "file-text",
  "cbt": "file-text",
  "cbtc": "file-text",
  "cbz": "file-text",
  "cc": "file-code",
  "cdr": "image",
  "cpp": "file-code",
  "css": "file-code",
  "csv": "file-excel-2",
  "cvbdl": "file-text",
  "c": "file-code",
  "c++": "file-code",
  "deb": "file-zip",
  "doc": "file-word-2",
  "docm": "file-word-2",
  "docx": "file-word-2",
  "dot": "file-word-2",
  "dotx": "file-word-2",
  "eml": "file-text",
  "eot": "image",
  "epub": "file-text",
  "eps": "image",
  "gz": "file-zip",
  "gzip": "file-zip",
  "h": "file-code",
  "hh": "file-code",
  "hpp": "file-code",
  "html": "file-code",
  "htm": "file-code",
  "java": "file-code",
  "js": "file-code",
  "json": "file-code",
  "jpeg": "image",
  "jpg": "image",
  "lwp": "file-word-2",
  "mdb": "file",
  "odp": "file-ppt-2",
  "ods": "file-excel-2",
  "odt": "file-word-2",
  "one": "file-word-2",
  "otf": "image",
  "pdf": "file-pdf",
  "pfb": "image",
  "php": "file-code",
  "pl": "file-code",
  "png": "image",
  "pot": "file-ppt-2",
  "potm": "file-ppt-2",
  "potx": "file-ppt-2",
  "ppa": "file-ppt-2",
  "ppam": "file-ppt-2",
  "pps": "file-ppt-2",
  "ppsm": "file-ppt-2",
  "ppsx": "file-ppt-2",
  "ppt": "file-ppt-2",
  "pptm": "file-ppt-2",
  "pptx": "file-ppt-2",
  "ps": "image",
  "psd": "image",
  "py": "file-code",
  "rar": "file-zip",
  "rss": "file",
  "sh-lib": "file-code",
  "sh": "file-code",
  "swf": "file",
  "tar": "file-zip",
  "tar.bz2": "file-zip",
  "tar.gz": "file-zip",
  "tbz2": "file-zip",
  "tex": "file-text",
  "tgz": "file-zip",
  "ttf": "image",
  "vsd": "file-word-2",
  "woff": "image",
  "wpd": "file-word-2",
  "xcf": "image",
  "xla": "file-excel-2",
  "xlam": "file-excel-2",
  "xls": "file-excel-2",
  "xlsb": "file-excel-2",
  "xlsm": "file-excel-2",
  "xlsx": "file-excel-2",
  "xlt": "file-excel-2",
  "xltm": "file-excel-2",
  "xltx": "file-excel-2",
  "xml": "file-code",
  "yaml": "file-code",
  "yml": "file-code",
  "zip": "file-zip"
}
