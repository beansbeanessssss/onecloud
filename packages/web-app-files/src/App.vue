<template>
  <main id="files" class="oc-flex oc-height-1-1">
    <div v-if="dragareaEnabled" class="dragarea" />
    <router-view tabindex="0" class="oc-width-expand" />
  </main>
</template>
<script lang="ts">
import { defineComponent, watch } from '@vue/composition-api'
import { Route } from 'vue-router'
import { useRoute, useStore } from 'web-pkg/src/composables'
import { bus } from 'web-pkg/src/instance'

export default defineComponent({
  setup() {
    const store = useStore<any>()
    watch(useRoute(), (to: Route, from?: Route) => {
      store.dispatch('Files/resetFileSelection')
    })
  },
  data: () => ({
    dragareaEnabled: false
  }),
  created() {
    const dragOver = bus.subscribe('drag-over', this.onDragOver)
    const dragOut = bus.subscribe('drag-out', this.hideDropzone)
    const drop = bus.subscribe('drop', this.hideDropzone)

    this.$on('beforeDestroy', () => {
      bus.unsubscribe('drag-over', dragOver)
      bus.unsubscribe('drag-out', dragOut)
      bus.unsubscribe('drop', drop)
    })
  },

  methods: {
    hideDropzone() {
      this.dragareaEnabled = false
    },
    onDragOver(event) {
      this.dragareaEnabled = (event.dataTransfer.types || []).some((e) => e === 'Files')
    }
  }
})
</script>

<style lang="scss" scoped>
main {
  max-height: 100%;
}

.dragarea {
  background-color: rgba(60, 130, 225, 0.21);
  pointer-events: none;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  position: absolute;
  z-index: 9;
  border-radius: 14px;
  border: 2px dashed var(--oc-color-swatch-primary-muted);
}

#files {
  position: relative;
}
</style>
