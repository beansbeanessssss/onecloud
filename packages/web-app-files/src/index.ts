import App from './App.vue'
import Favorites from './views/Favorites.vue'
import FilesDrop from './views/FilesDrop.vue'
import SharedWithMe from './views/shares/SharedWithMe.vue'
import SharedWithOthers from './views/shares/SharedWithOthers.vue'
import SharedViaLink from './views/shares/SharedViaLink.vue'
import SpaceDriveResolver from './views/spaces/DriveResolver.vue'
import SpaceProjects from './views/spaces/Projects.vue'
import TrashOverview from './views/trash/Overview.vue'
import translations from '../l10n/translations.json'
import { defineWebApplication, quickActions } from '@ownclouders/web-pkg'
import store from './store'
import { extensions } from './extensions'
import fileSideBars from './fileSideBars'
import { buildRoutes } from '@ownclouders/web-pkg'
import { AppNavigationItem } from '@ownclouders/web-pkg'

// dirty: importing view from other extension within project
import SearchResults from '../../web-app-search/src/views/List.vue'
import {
  SpaceResource,
  isPersonalSpaceResource,
  isShareSpaceResource
} from '@ownclouders/web-client/src/helpers'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appInfo = {
  name: $gettext('Files'),
  id: 'files',
  icon: 'resource-type-folder',
  color: 'var(--oc-color-swatch-primary-muted)',
  isFileEditor: false,
  extensions: [],
  fileSideBars
}
export const navItems = (context): AppNavigationItem[] => {
  return [
    {
      name(capabilities) {
        return capabilities.spaces?.enabled ? $gettext('Personal') : $gettext('All files')
      },
      icon: appInfo.icon,
      route: {
        path: `/${appInfo.id}/spaces/personal`
      },
      isActive: () => {
        return (
          !context.$store.getters['runtime/spaces/currentSpace'] ||
          context.$store.getters['runtime/spaces/currentSpace']?.isOwner(
            context.$store.getters.user
          )
        )
      },
      enabled(capabilities) {
        if (!capabilities.spaces?.enabled) {
          return true
        }
        return !!context?.$store?.getters['runtime/spaces/spaces'].find(
          (drive) => isPersonalSpaceResource(drive) && drive.isOwner(context.$store.getters.user)
        )
      }
    },
    {
      name: $gettext('Favorites'),
      icon: 'star',
      route: {
        path: `/${appInfo.id}/favorites`
      },
      enabled(capabilities) {
        return capabilities.files?.favorites
      }
    },
    {
      name: $gettext('Shares'),
      icon: 'share-forward',
      route: {
        path: `/${appInfo.id}/shares`
      },
      isActive: () => {
        const space = context.$store.getters['runtime/spaces/currentSpace'] as SpaceResource
        // last check is when fullShareOwnerPaths is enabled
        return !space || isShareSpaceResource(space) || !space?.isOwner(context.$store.getters.user)
      },
      activeFor: [
        { path: `/${appInfo.id}/spaces/share` },
        { path: `/${appInfo.id}/spaces/personal` }
      ],
      enabled(capabilities) {
        return capabilities.files_sharing?.api_enabled !== false
      }
    },
    {
      name: $gettext('Spaces'),
      icon: 'layout-grid',
      route: {
        path: `/${appInfo.id}/spaces/projects`
      },
      activeFor: [{ path: `/${appInfo.id}/spaces/project` }],
      enabled(capabilities) {
        return capabilities.spaces?.projects
      }
    },
    {
      name: $gettext('Deleted files'),
      icon: 'delete-bin-5',
      route: {
        path: `/${appInfo.id}/trash/overview`
      },
      activeFor: [{ path: `/${appInfo.id}/trash` }],
      enabled(capabilities) {
        return capabilities.dav?.trashbin === '1.0' && capabilities.files?.undelete
      }
    }
  ]
}

export default defineWebApplication({
  setup(args) {
    return {
      appInfo,
      store,
      routes: buildRoutes({
        App,
        Favorites,
        FilesDrop,
        SearchResults,
        Shares: {
          SharedViaLink,
          SharedWithMe,
          SharedWithOthers
        },
        Spaces: {
          DriveResolver: SpaceDriveResolver,
          Projects: SpaceProjects
        },
        Trash: {
          Overview: TrashOverview
        }
      }),
      navItems,
      quickActions,
      translations,
      extensions: extensions(args)
    }
  }
})
