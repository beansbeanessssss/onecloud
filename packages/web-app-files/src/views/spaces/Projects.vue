<template>
  <div class="oc-flex">
    <files-view-wrapper>
      <app-bar
        :breadcrumbs="breadcrumbs"
        :has-view-options="false"
        :has-sidebar-toggle="true"
        :show-actions-on-selection="true"
        :side-bar-open="sideBarOpen"
      >
        <template #actions>
          <create-space v-if="hasCreatePermission" />
        </template>
      </app-bar>
      <app-loading-spinner v-if="areResourcesLoading" />
      <template v-else>
        <no-content-message
          v-if="!spaces.length"
          id="files-spaces-empty"
          class="files-empty"
          icon="layout-grid"
        >
          <template #message>
            <span v-translate>You don't have access to any spaces</span>
          </template>
        </no-content-message>
        <div v-else class="spaces-list oc-px-m oc-mt-l">
          <resource-tiles :data="spaces" @contextmenuClicked="showContextMenu">
            <template #image="{ item }">
              <img
                v-if="imageContentObject[item.id]"
                class="space-image tile-image oc-rounded-top"
                :src="imageContentObject[item.id]['data']"
                alt=""
              />
            </template>
            <template #actions="{ item }">
              <div>
                <oc-button
                  v-oc-tooltip="showSpaceMemberLabel"
                  :aria-label="showSpaceMemberLabel"
                  appearance="raw"
                  @click="openSidebarSharePanel(item)"
                >
                  <oc-icon name="group" fill-type="line" />
                </oc-button>
              </div>
              <div>
                <oc-button
                  :id="`space-context-btn-${item.getDomSelector()}`"
                  :ref="`spaceContextBtn-${item.getDomSelector()}`"
                  v-oc-tooltip="contextMenuLabel"
                  :aria-label="contextMenuLabel"
                  appearance="raw"
                  @click.stop.prevent="resetDropPosition($event, item)"
                >
                  <oc-icon name="more-2" />
                </oc-button>
                <oc-drop
                  :ref="`spaceContextDrop-${item.getDomSelector()}`"
                  :drop-id="`space-context-drop-${item.getDomSelector()}`"
                  close-on-click
                  mode="manual"
                  :options="{ delayHide: 0 }"
                  padding-size="small"
                >
                  <space-context-actions :space="item" :items="[item]" />
                </oc-drop>
              </div>
            </template>
          </resource-tiles>
        </div>
      </template>
    </files-view-wrapper>
    <side-bar :open="sideBarOpen" :active-panel="sideBarActivePanel" :space="highlightedFile" />
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, unref } from 'vue'
import { useTask } from 'vue-concurrency'
import { mapMutations, mapGetters } from 'vuex'

import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'

import AppBar from '../../components/AppBar/AppBar.vue'
import CreateSpace from '../../components/AppBar/CreateSpace.vue'
import { useAccessToken, useStore, useGraphClient } from 'web-pkg/src/composables'
import { loadPreview } from 'web-pkg/src/helpers/preview'
import { ImageDimension } from '../../constants'
import SpaceContextActions from '../../components/Spaces/SpaceContextActions.vue'
import { configurationManager } from 'web-pkg/src/configuration'
import { isProjectSpaceResource, SpaceResource } from 'web-client/src/helpers'
import SideBar from '../../components/SideBar/SideBar.vue'
import FilesViewWrapper from '../../components/FilesViewWrapper.vue'
import ResourceTiles from '../../components/FilesList/ResourceTiles.vue'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { SideBarEventTopics, useSideBar } from '../../composables/sideBar'
import { WebDAV } from 'web-client/src/webdav'
import { useScrollTo } from 'web-app-files/src/composables/scrollTo'

export default defineComponent({
  components: {
    ResourceTiles,
    FilesViewWrapper,
    SideBar,
    AppBar,
    AppLoadingSpinner,
    CreateSpace,
    NoContentMessage,
    SpaceContextActions
  },
  setup() {
    const store = useStore()
    const spaces = computed(
      () =>
        store.getters['runtime/spaces/spaces']
          .filter((s) => isProjectSpaceResource(s))
          .sort((a, b) => a.name.localeCompare(b.name)) || []
    )
    const accessToken = useAccessToken({ store })
    const { graphClient } = useGraphClient()

    const loadResourcesTask = useTask(function* () {
      store.commit('Files/CLEAR_FILES_SEARCHED')
      store.commit('Files/CLEAR_CURRENT_FILES_LIST')
      yield store.dispatch('runtime/spaces/reloadProjectSpaces', {
        graphClient: unref(graphClient)
      })
      store.commit('Files/LOAD_FILES', { currentFolder: null, files: unref(spaces) })
    })

    const areResourcesLoading = computed(() => {
      return loadResourcesTask.isRunning || !loadResourcesTask.last
    })

    return {
      ...useSideBar(),
      ...useScrollTo(),
      spaces,
      graphClient,
      loadResourcesTask,
      areResourcesLoading,
      accessToken
    }
  },
  data: function () {
    return {
      imageContentObject: {}
    }
  },
  computed: {
    ...mapGetters(['user']),
    ...mapGetters('Files', ['highlightedFile']),
    breadcrumbs() {
      return [
        {
          text: this.$gettext('Spaces'),
          onClick: () => this.loadResourcesTask.perform()
        }
      ]
    },
    showSpaceMemberLabel() {
      return this.$gettext('Show members')
    },
    contextMenuLabel() {
      return this.$gettext('Show context menu')
    },
    hasCreatePermission() {
      return this.$permissionManager.hasSpaceManagement()
    }
  },
  watch: {
    spaces: {
      handler: async function (val) {
        if (!val) {
          return
        }

        for (const space of this.spaces as SpaceResource[]) {
          if (!space.spaceImageData) {
            continue
          }

          if (this.imageContentObject[space.id]?.fileId === space.spaceImageData?.id) {
            continue
          }

          const decodedUri = decodeURI(space.spaceImageData.webDavUrl)
          const webDavPathComponents = decodedUri.split('/')
          const idComponent = webDavPathComponents.find((c) => c.startsWith(`${space.id}`))
          if (!idComponent) {
            return
          }
          const path = webDavPathComponents
            .slice(webDavPathComponents.indexOf(idComponent) + 1)
            .join('/')

          const resource = await (this.$clientService.webdav as WebDAV).getFileInfo(space, { path })
          loadPreview({
            resource,
            isPublic: false,
            dimensions: ImageDimension.Preview,
            server: configurationManager.serverUrl,
            userId: this.user.id,
            token: this.accessToken
          }).then((imageBlob) => {
            this.$set(this.imageContentObject, space.id, {
              fileId: space.spaceImageData.id,
              data: imageBlob
            })
          })
        }
      },
      deep: true,
      immediate: true
    }
  },
  async created() {
    await this.loadResourcesTask.perform()
    this.scrollToResourceFromRoute(this.spaces)
  },
  methods: {
    ...mapMutations('Files', ['SET_FILE_SELECTION']),

    resetDropPosition(event, item) {
      const drop = this.$refs[`spaceContextDrop-${item.getDomSelector()}`]

      if (drop === undefined) {
        return
      }
      this.displayPositionedDropdown(drop, event, item)
    },
    showContextMenu(_row, event, item) {
      event.preventDefault()
      const drop = this.$refs[`spaceContextDrop-${item.getDomSelector()}`]

      this.displayPositionedDropdown(drop, event, item)
    },
    displayPositionedDropdown(dropdown, event, item) {
      const contextMenuButtonPos =
        this.$refs[`spaceContextBtn-${item.getDomSelector()}`].$el.getBoundingClientRect()

      if (!dropdown || !contextMenuButtonPos) {
        return
      }

      dropdown.tippy.setProps({
        getReferenceClientRect: () => ({
          width: 0,
          height: 0,
          top: event.clientY,
          bottom: event.clientY,
          left: event.type === 'contextmenu' ? event.clientX : contextMenuButtonPos.x,
          right: event.type === 'contextmenu' ? event.clientX : contextMenuButtonPos.x
        })
      })
      dropdown.show()
    },
    openSidebarSharePanel(space: SpaceResource) {
      this.SET_FILE_SELECTION([space])
      eventBus.publish(SideBarEventTopics.openWithPanel, 'space-share')
    }
  }
})
</script>

<style lang="scss">
#files-spaces-empty {
  height: 75vh;
}

.state-trashed {
  .tile-image,
  .tile-default-image > svg {
    filter: grayscale(100%);
    opacity: 80%;
  }
}

.space-image {
  width: 100%;
  aspect-ratio: 16/9;
  object-fit: cover;
}
</style>
