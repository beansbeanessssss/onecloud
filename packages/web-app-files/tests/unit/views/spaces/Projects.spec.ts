import Projects from '../../../../src/views/spaces/Projects.vue'
import { mockDeep } from 'jest-mock-extended'
import { SpaceResource } from 'web-client/src/helpers'
import {
  createStore,
  defaultPlugins,
  mount,
  defaultStoreMockOptions,
  defaultComponentMocks,
  defaultStubs
} from 'web-test-helpers'

describe('Projects view', () => {
  it('appBar always present', () => {
    const { wrapper } = getMountedWrapper()
    expect(wrapper.find('app-bar-stub').exists()).toBeTruthy()
  })
  it('sideBar always present', () => {
    const { wrapper } = getMountedWrapper()
    expect(wrapper.find('side-bar-stub').exists()).toBeTruthy()
  })
  describe('different files view states', () => {
    it('shows the loading spinner during loading', () => {
      const { wrapper } = getMountedWrapper()
      expect(wrapper.find('oc-spinner-stub').exists()).toBeTruthy()
    })
    it('shows the no-content-message after loading', async () => {
      const { wrapper } = getMountedWrapper()
      await wrapper.vm.loadResourcesTask.last
      expect(wrapper.find('oc-spinner-stub').exists()).toBeFalsy()
      expect(wrapper.find('.no-content-message').exists()).toBeTruthy()
    })
    it('lists all available project spaces', async () => {
      const spaces = [
        mockDeep<SpaceResource>({
          id: '1',
          name: 'Some space',
          driveType: 'project',
          description: 'desc'
        }),
        mockDeep<SpaceResource>({
          id: '2',
          name: 'Some other space',
          driveType: 'project',
          description: 'desc'
        })
      ]
      const { wrapper } = getMountedWrapper({ spaces })
      await wrapper.vm.loadResourcesTask.last
      // "space" is undefined for "space-context-actions", seems to be a bug because it's definitely not
      // {{ space }} -> undefined, {{ space.id }} -> "1"
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.find('.no-content-message').exists()).toBeFalsy()
      expect(wrapper.find('.spaces-list').exists()).toBeTruthy()
      expect(wrapper.findAll('.oc-tiles-item').length).toEqual(spaces.length)
    })
  })
})

function getMountedWrapper({ mocks = {}, spaces = [] } = {}) {
  const defaultMocks = {
    ...defaultComponentMocks({
      currentRoute: { name: 'files-spaces-projects' }
    }),
    ...(mocks && mocks)
  }
  const storeOptions = { ...defaultStoreMockOptions }
  storeOptions.modules.runtime.modules.spaces.getters.spaces = jest.fn(() => spaces)
  const store = createStore(storeOptions)
  return {
    mocks: defaultMocks,
    storeOptions,
    wrapper: mount(Projects, {
      global: {
        plugins: [...defaultPlugins(), store],
        mocks: defaultMocks,
        stubs: { ...defaultStubs, 'space-context-actions': true }
      }
    })
  }
}
