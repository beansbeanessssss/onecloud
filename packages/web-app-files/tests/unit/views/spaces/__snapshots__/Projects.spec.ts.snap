// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects view different files view states lists all available project spaces 1`] = `
<div class="oc-flex">
  <div class="files-view-wrapper oc-width-expand">
    <div id="files-view">
      <app-bar-stub breadcrumbs="[object Object]" breadcrumbscontextactionsitems="" hassidebartoggle="true" showactionsonselection="true"></app-bar-stub>
      <div class="spaces-list oc-px-m oc-mt-l">
        <div>
          <ul class="oc-list oc-my-rm oc-mx-rm oc-tiles oc-flex">
            <li class="oc-tiles-item">
              <oc-tile-stub resource="undefined" resourceroute="[object Object]"></oc-tile-stub>
            </li>
            <li class="oc-tiles-item">
              <oc-tile-stub resource="undefined" resourceroute="[object Object]"></oc-tile-stub>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <side-bar-stub></side-bar-stub>
</div>
`;
